File: Documentation/5.4/Samples/java/src/test/java/net/ravendb/Indexes/Querying/MoreLikeThis.java
package net.ravendb.Indexes.Querying;

import net.ravendb.client.documents.DocumentStore;
import net.ravendb.client.documents.IDocumentStore;
import net.ravendb.client.documents.indexes.AbstractIndexCreationTask;
import net.ravendb.client.documents.indexes.FieldStorage;
import net.ravendb.client.documents.queries.moreLikeThis.MoreLikeThisOptions;
import net.ravendb.client.documents.queries.moreLikeThis.MoreLikeThisStopWords;
import net.ravendb.client.documents.session.IDocumentSession;

import java.util.Arrays;
import java.util.List;

public class MoreLikeThis {

    //region more_like_this_4
    public class Article {
        private String id;
        private String name;
        private String articleBody;

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getArticleBody() {
            return articleBody;
        }

        public void setArticleBody(String articleBody) {
            this.articleBody = articleBody;
        }
    }

    public class Articles_ByArticleBody extends AbstractIndexCreationTask {
        public Articles_ByArticleBody() {
            map = "from doc in docs.articles " +
                "select new {" +
                "   doc.articleBody " +
                "}";

            store("articleBody", FieldStorage.YES);
            analyze("articleBody", "StandardAnalyzer");
        }
    }
    //endregion

    public MoreLikeThis() {
        try (IDocumentStore store = new DocumentStore()) {
            try (IDocumentSession session = store.openSession()) {
                {
                    //region more_like_this_1
                    List<Article> articles = session
                        .query(Article.class, Articles_ByArticleBody.class)
                        .moreLikeThis(builder -> builder.usingDocument(x -> x.whereEquals("id()", "articles/1")))
                        .toList();
                    //endregion
                }

                {
                    //region more_like_this_2
                    MoreLikeThisOptions options = new MoreLikeThisOptions();
                    options.setFields(new String[]{ "articleBody" });
                    List<Article> articles = session
                        .query(Article.class, Articles_ByArticleBody.class)
                        .moreLikeThis(builder -> builder
                            .usingDocument(x -> x.whereEquals("id()", "articles/1"))
                            .withOptions(options))
                        .toList();
                    //endregion
                }

                {
                    //region more_like_this_3
                    MoreLikeThisStopWords stopWords = new MoreLikeThisStopWords();
                    stopWords.setStopWords(Arrays.asList("I", "A", "Be"));
                    session.store(stopWords, "Config/Stopwords");
                    //endregion
                }

            }
        }
    }
}

File: Documentation/5.4/Samples/php/Indexes/Querying/MoreLikeThis.php
<?php

namespace RavenDB\Samples\Indexes\Querying;

use RavenDB\Documents\DocumentStore;
use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Documents\Indexes\FieldStorage;
use RavenDB\Documents\Queries\MoreLikeThis\MoreLikeThisOptions;
use RavenDB\Documents\Queries\MoreLikeThis\MoreLikeThisStopWords;

class MoreLikeThis
{
    public function Sample(): void
    {
        $store = new DocumentStore();
        try {
            $session = $store->openSession();
            try {
                # region more_like_this_1
                /** @var array<Article> $articles */
                $articles = $session
                    ->query(Article::class, Articles_ByArticleBody::class)
                    ->moreLikeThis(function($builder) {
                        return $builder
                            ->usingDocument(function($x) {
                                return $x->whereEquals("id()", "articles/1");
                            });
                    })
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region more_like_this_1_1
                /** @var array<Article> $articles */
                $articles = $session->advanced()
                    ->documentQuery(Article::class, Articles_ByArticleBody::class)
                    ->moreLikeThis(function($builder) {
                        return $builder
                            ->usingDocument(function($x) {
                                return $x->whereEquals("id()", "articles/1");
                            });
                    })
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region more_like_this_2
                /** @var array<Article> $articles */
                $articles = $session
                    ->query(Article::class, Articles_ByArticleBody::class)
                    ->moreLikeThis(function($builder) {
                        $mlt = new MoreLikeThisOptions();
                        $mlt->setFields(["ArticleBody"]);

                        return $builder
                            ->usingDocument(function($x) {
                                return $x->whereEquals("id()", "articles/1");
                            })
                            ->withOptions($mlt);
                    })
                    ->toList();
                # endregion

                # region more_like_this_3
                $mlt = new MoreLikeThisStopWords();
                $mlt->setId("Config/Stopwords");
                $mlt->setStopWords(["I", "A", "Be"]);
                $session->store($mlt);
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region more_like_this_2_1
                /** @var array<Article> $articles */
                $articles = $session->advanced()
                    ->documentQuery(Article::class, Articles_ByArticleBody::class)
                    ->moreLikeThis(function($builder) {
                        $mlt = new MoreLikeThisOptions();
                        $mlt->setFields(["ArticleBody"]);

                        return $builder
                            ->usingDocument(function($x) {
                                return $x->whereEquals("id()", "articles/1");
                            })
                            ->withOptions($mlt);
                    })
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}

# region more_like_this_4
class Article
{
    public ?string $id = null;
    public ?string $name = null;
    public ?string $articleBody = null;

    public function getId(): ?string
    {
        return $this->id;
    }

    public function setId(?string $id): void
    {
        $this->id = $id;
    }

    public function getName(): ?string
    {
        return $this->name;
    }

    public function setName(?string $name): void
    {
        $this->name = $name;
    }

    public function getArticleBody(): ?string
    {
        return $this->articleBody;
    }

    public function setArticleBody(?string $articleBody): void
    {
        $this->articleBody = $articleBody;
    }
}

class Articles_ByArticleBody extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->map = "from doc in docs.Articles select { doc.ArticleBody }";

        $this->store("ArticleBody", FieldStorage::yes());
        $this->analyze("ArticleBody", "StandardAnalyzer");
    }
}
# endregion

